aggregator/election_mgr.go

```diff
@@ -0,0 +1,545 @@
+		// to follower transition is no longer valid.
+		if !(state.id == currState.id && state.state == currState.state) {
+			mgr.metrics.verifyPendingChangeStale.Inc(1)
+			mgr.goalStateLock.Unlock()
+			continue
+		}
+		if !knownLeader {
+			mgr.metrics.verifyNoKnownLeader.Inc(1)
+			mgr.goalStateLock.Unlock()
+			continue
+		}
+		if isLeader {
+			mgr.setGoalStateWithLock(LeaderState)
+		} else {
+			mgr.setGoalStateWithLock(FollowerState)
+		}
+		mgr.goalStateLock.Unlock()
+	}
+}
+
+func (mgr *electionManager) resign(action resignAction) {
+	mgr.RLock()
+	if mgr.state != electionManagerOpen {
+		action.errCh <- errElectionManagerNotOpenOrClosed
+		mgr.RUnlock()
+		return
+	}
+	mgr.RUnlock()
+
+	if mgr.ElectionState() == FollowerState {
+		mgr.metrics.followerResign.Inc(1)
+		action.errCh <- nil
+		return
+	}
+
+	if err := mgr.leaderService.Resign(mgr.electionKey); err != nil {
+		action.errCh <- err
+	}
+}
+
+func (mgr *electionManager) campaignLoop() {
+	var campaignStatusCh <-chan campaign.Status
+	notDone := func(int) bool {
+		select {
+		case <-mgr.doneCh:
+			return false
+		default:
+			return true
+		}
+	}
+
+	for {
+		if campaignStatusCh == nil {
+			// NB(xichen): campaign retrier retries forever until either the Campaign call succeeds,
+			// or the election manager has resigned.
+			if err := mgr.campaignRetrier.AttemptWhile(notDone, func() error {
+				var err error
+				campaignStatusCh, err = mgr.leaderService.Campaign(mgr.electionKey, mgr.campaignOpts)
+				if err == nil {
+					return nil
+				}
+				mgr.metrics.campaignCreateErrors.Inc(1)
+				mgr.logError("error creating campaign", err)
+				return err
+			}); err == xretry.ErrWhileConditionFalse {
+				return
+			}
+		}
+
+		var (
+			campaignStatus campaign.Status
+			ok             bool
+		)
+		select {
+		case campaignStatus, ok = <-campaignStatusCh:
+			// If the campaign status channel is closed, this is either because session has expired,
+			// or we have resigned from the campaign, or there are issues with the underlying etcd
+			// cluster, in which case we back off a little and restart the campaign.
+			if !ok {
+				campaignStatusCh = nil
+				mgr.sleepFn(backOffOnResignOrElectionError)
+				continue
+			}
+		case <-mgr.doneCh:
+			return
+		}
+		mgr.processCampaignUpdate(campaignStatus)
+	}
+}
+
+func (mgr *electionManager) processCampaignUpdate(campaignStatus campaign.Status) {
+	if campaignStatus.State == campaign.Error {
+		mgr.metrics.campaignErrors.Inc(1)
+		mgr.logError("error campaigning", campaignStatus.Err)
+		return
+	}
+
+	newState, err := newElectionState(campaignStatus.State)
+	if err != nil {
+		mgr.metrics.campaignUnknownState.Inc(1)
+		mgr.logError("unknown campaign state", err)
+		return
+	}
+
+	mgr.goalStateLock.Lock()
+	mgr.setGoalStateWithLock(newState)
+	mgr.goalStateLock.Unlock()
+}
+
+func (mgr *electionManager) setGoalStateWithLock(newState ElectionState) {
+	newGoalStateID := mgr.goalStateID
```

Comment:
Nit: Is it weird calling this newGoalStateID since we increment it right after, maybe just goalStateID? Also, is it weird that mgr.goalStateID is actually 1 more than the current goalStateID?